enclave/core/CMakeLists.txt

```diff
@@ -275,5 +278,10 @@ elseif(OE_TRUSTZONE)
         --as-needed)
 endif()
 
+# Ensure all moodule initialization functions are linked into the
+# enclave binary regardless of whether they are used explicitly
+target_link_libraries(oelibc INTERFACE
+    -Wl,-u,mod_init*)
```

Comment:
Consider documenting this somewhere as a required linker flag for enclave which use OE modules. Ideally there is a cleaner way to do this, but I haven't found one yet.

Reply:
This doesn't actually work. It only works if the full symbol name is given.

Reply:
@radhikaj @CodeMonkeyLeet @anakrish 

One way to force the module init function to link into the enclave binary is to use this linker hack (`-u <symbol_name`).

Another way is to reference the symbol in the main executable. This doesn't need to be via a function call. We could add a macro `OE_MODULE_INCLUDE(<mod_name>);` that can be added anywhere in the enclave code. This will ensure the initialization function is linked.

The only problem with this is that because modules are really just .a libraries, if the developer forgets to add this line, they can still call functions in the module (such as libc functions), the module will just not have been initialized yet. That said, the same is true of if the developer forgets to add the `-u` linker flag.

I don't really like the linker hack but it does work. Wanted to get some thoughts on which approach people prefer.

Reply:
@jhand2 I'd prototyped a similar approach earlier: https://github.com/openenclave/openenclave/compare/anakrish/feature/initializers
I am not sure whether it would work given that we now aggressively remove sections.

I like your idea of using constructor attributes and priorities. If we can make that work, then the compiler/linker takes care of most of the work for us.

Reply:
I think the problem is that the constructor attribute isn't designed to allow you to insert calls prior to C++ constructors. I think it is better to just implement priorities ourselves. Its pretty easy, you can just have initialization stages with a different section for each stage. So your sections would be: `.oeinit.early`, `.oeinit.default`, `.oeinit.late`, etc. Either way, we still run into the linker optimization problem.

The linux kernel actually faces a similar problem with kernel drivers. Drivers can either be dynamically loadable (.ko shared object files) or compiled into the kernel statically. They address this problem by using a linker script that marks their .init section with a `KEEP()`.

I think a linker script is actually the correct way to do this. The problem is that Linux can handle this a little more cleanly because they control the linking step. Technically we don't, so we would have to provide a linker script and get the enclave developer to pass the linker script to gcc/clang.